README.md
# Docker implementation of the Station Demand Forecasting Tool

A quick and easy install of the Station Demand Forecasting Tool on any modern operating system, using Docker.

It is assumed that you have Docker installed and running on a host computer. This could be a virtual machine in a cloud environment, such as Google Cloud, AWS, Microsoft Azure, or Digital Ocean. For testing purposes or a very simple model run (a single station) you could use your own computer, provided it has sufficient resources available.

We have had good results using a DigitalOcean CPU-Optimized virtual machine with 16 or 32 CPUs and using the Docker image created by DigitalOcean that's available in the Marketplace. DigitalOcean virtual machines are charged by the second (*whether running or not*), and the hourly charge for the CPU-Optimized 32 CPU VM is just under $1.00. For new users, [this link](https://m.do.co/c/86d69abc23ef) will get you a $100 (60-day) credit.

The Docker implementation consists of two containers:

1. An instance of Rstudio Server with all the required packages and dependencies installed, including the [sdft R package](https://github.com/station-demand-forecasting-tool/sdft).

2. An instance of PostgreSQL server with the PostGIS and pgRouting extensions installed, and all the database tables required by the model.

Images for these containers are available via the Docker Hub. There is no need to clone this repository or generate the images yourself.

## Setup Instructions

1. Copy the [docker-compose.yml](https://raw.githubusercontent.com/station-demand-forecasting-tool/sdft-docker/master/docker-compose.yml) file from the repository to the host computer (place it in a directory called `sdft-docker`). 

2. Edit `docker-compose.yml` and replace the two instances of `your_password` with a password of your choice. This will set the postgres user password and the rstudio user password.

3. Edit `docker-compose.yml` and amend the entry:  

   ```yaml
   volumes:
      - c:/sdft:/home/rstudio
   ```

   You should replace `c:/sdft` with the path to a suitable directory on the ***host*** computer. In the example above, it is a windows host and the location is `c:/sdft`. This folder will be used to read the input files for a model job. It will also be used to write the outputs from a model job. The location needs to be readable and writeable by the 'rstudio' user in the Rstudio container. On Linux, you could create a new directory at root level and change its ownership to user id 1000 (this is the id of the 'rstudio' user):
   
   ```bash
   mkdir /sdft/
   chown 1000 /sdft/
   ```
   and then amend `docker-compose.yml` as follows:
   
   ```yaml
   volumes:
      - /sdft:/home/rstudio
   ```

3. You may want to check for the most recent version of the images on Docker Hub. You can check using these links:  
   - [https://hub.docker.com/r/sdft/sdft-ui/tags](https://hub.docker.com/r/sdft/sdft-ui/tags)
   - [https://hub.docker.com/r/sdft/sdft-db/tags](https://hub.docker.com/r/sdft/sdft-db/tags)    

   Then amend the image tags in `docker-compose.yaml` as required.

3. In terminal or command prompt change to your `sdft-docker` folder.

4. Run: `docker-compose up -d`

5. The images will be downloaded from Docker Hub and inflated; if they are not already stored locally.

6. On successful completion you should see the following:  

   ```bash
   Creating network "sdft-docker_postgresql" with driver "bridge"
   Creating sdft-db ... done
   Creating sdft-ui ... done
   ```

7. You can now connect to the Rstudio server at [http://locahost:8787](http://localhost:8787). Logon with the user `rstudio` and the password you entered in `docker-compose.yml` earlier. If you are installing on a cloud-based VM then you will usually want to configure a tunnel in your SSH client to forward port 8787 on your local computer to port 8787 on your VM (By default, the DigitalOcean Docker image will allow external access from any Internet host to http://yourdropletipaddress:8787). 

8. On the PostgreSQL container the database will now be created and the required tables and indexes generated. This process may take some time (perhaps 30-60 minutes). Due to the size of the database once fully generated (around 10GB), this is the most practical method of distribution. This process is only carried out the first time the sdft-db container is built. You can stop and start this container and the database will be preserved.

9. To check on progress, you can attach to the container by running: `docker attach  --sig-proxy=false sdft-db` from the command prompt or terminal. The `--sig-proxy=false` option allows you to use `ctrl-c` to detach from the container without stopping it. Once the database initialisation is complete, you will see the following (assuming you're still attached to the container):

   ```bash
   PostgreSQL init process complete; ready for start up.

    2020-09-14 13:44:12.985 UTC [1] LOG:  starting PostgreSQL 12.4 (Debian 12.4-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
    2020-09-14 13:44:12.985 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
    2020-09-14 13:44:12.985 UTC [1] LOG:  listening on IPv6 address "::", port 5432
    2020-09-14 13:44:12.991 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
    2020-09-14 13:44:13.012 UTC [1457] LOG:  database system was shut down at 2020-09-14 13:44:12 UTC
    2020-09-14 13:44:13.018 UTC [1] LOG:  database system is ready to accept connections
    
   ```


## Running a job

This provides brief instructions on running a job in testing mode. Please consult the [documentation](https://www.stationdemand.org.uk) for further information.

1. In your browser connect to the Rstudio server at [http://localhost:8787](http://localhost:8787).

2. Copy the example job submission CSV files from the SDFT R package to `/home/rstudio/input` by running:
   
   ```r
   dir.create("/home/rstudio/input")
   files <- list.files(file.path(system.file(package = "sdft"), "example_input"), full.names = TRUE)
   file.copy(from=files, to="/home/rstudio/input")
   ```

3. The example files will run a job in testing mode. The testing mode only considers a very small catchment area for a proposed station to speed up processing.

4. Edit `config.csv` and set the number of processor cores to be used for the job. As a minimum, 4 are required. Save the file.

5. Provide your password for the postgres user, using `key_set()` from the keyring package: 
   
   ```r
   library(keyring)
   key_set("postgres")
   ```
   
   You will first be asked to provide a password for the 'system' keyring. This is a new password and you can provide whatever you like here. You will then be prompted for the password to set for the "postgres" service. You should enter the password that you set earlier in the `docker-compose.yml` file.
   
6. You are now ready to submit the test job:

   ```r
   library(sdft)
   sdft_submit(dbhost = "sdft-db", dirpath = "/home/rstudio")
   ```
   
7. A log file, `sdr.log`, will be generated in the `/home/rstudio/output` folder and will be updated while the job runs.

8. When the job is complete, several files are created in a subfolder of the output folder. The subfolder takes the name of the job as specified in the `config.csv` file. The set of files, assuming you used the example input as provided, are as follows:
   
   * `station_forecast.csv` contains the model forecast (remember, this will not be a correct forecast when testing mode is used)
   * `exogenous.csv` contains a summary of the input exogenous data
   * `helst1_catchment.geojson` contains the postcode centroids for the station's probabilistic catchment, along with the probability of the station being chosen as an attribute to each centroid. A interpolated surface can be generated using QGIS to visualise the catchment.
   * `sdr.log` contains information on the job run. The level of detail will depend on the `loglevel` set in the `config.csv` file. This is set to DEBUG for the example test run.
   
## Getting help

To report an issue, bug, or to make a feature request, please [create an issue in this repository](https://github.com/station-demand-forecasting-tool/sdft-docker/issues).

## Further information

Please see the [full documentation](https://www.stationdemand.org.uk).


NEWS.md
# sdft-docker 1.1.1

## Fixes

* Issues with /db/setup.sh:
  - Windows CR LF ^M error
  - added IF NOT EXISTS to all index creation statements.
* Issues with /db/data/stations.stations_load.sql:
  - error in create table command
  - creating index idx_stations_location_geom which already exists causes
  error

# sdft-docker 1.1.0

## Enhancements

* Virtual nodes are now created every 500m on edges > 1km. Previously they were 
created every 1km on edges < 1.5km.

## Data

* OS OpenRoads updated - 10/03/2021
* stations_upload updated with latest station openings
* stations.txt updated with latest station openings

## Fixes

* Issue passing db password to clusters in prepare_stations.R

## Other

* /db docker file - change to version argument for LFS downloads
* Additional comments added to docker-compose
* Now using rocker/rstudio:4.1.0 for ui container


# sdft-docker 1.0

Initial release


db/setup.sh
#!/bin/bash

# For local image generation only:
# if the stations variable is set to "create" then the station table will be generated from the stations.csv file
# and, once the database has been built on the sdft-db container, the
# prepare_stations.R script must be run from the sdft-ui container to generate the service areas for each station.
# The stations table can then be exported to a SQL file and used to generate a new public container image.
# We cannot use a shapefile for this export/import as there are multiple geometries in the stations table.
# stations="create"
stations="load"

set -e

psql --username "postgres" -f /home/sql/setup.sql

#prepare the tables don't load data -p  Prepare mode, only creates the table.
#force non-multi and set the geometry column name to the_geom, dbf is in latin1 encoding


shp2pgsql -s 27700 -g the_geom -S -p /home/openroads/data/HP_RoadLink.shp openroads.roadlinks | psql -U postgres -d dafni


F1=/home/openroads/data/*RoadLink.shp
for f in $F1
do
	shp2pgsql -s 27700 -g the_geom -S -a -D $f "openroads"."roadlinks" | psql -U postgres -d dafni
done


shp2pgsql -s 27700 -g the_geom -S -p /home/openroads/data/HP_RoadNode.shp openroads.roadnodes | psql -U postgres -d dafni


F2=/home/openroads/data/*RoadNode.shp
for g in $F2
do
	shp2pgsql -s 27700 -g the_geom -S -a -D $g "openroads"."roadnodes" | psql -U postgres -d dafni
done

psql --username "postgres" -d dafni -f /home/sql/openroads.sql

psql --username "postgres" -d dafni -f /home/sql/disconnected.sql

psql --username "postgres" -d dafni -f /home/sql/virtualnodes.sql


# load pc_pop_2011 shapefile

f=/home/data/population/pc_pop_2011.shp

shp2pgsql -s 27700 -g geom -S -D $f "data"."pc_pop_2011" | psql -U postgres -d dafni

psql --username "postgres" -d dafni -c 'alter table data.pc_pop_2011 drop column id;'
psql --username "postgres" -d dafni -c 'alter table data.pc_pop_2011 rename column gid to id;'
psql --username "postgres" -d dafni -c 'create index IF NOT EXISTS idx_pc_pop_2011_geom on data.pc_pop_2011 using gist ( geom );'
psql --username "postgres" -d dafni -c 'create index IF NOT EXISTS idx_pc_pop_2011_postcode on data.pc_pop_2011 using btree ( postcode );'

# decision whether to load stations shapefile with service areas
# or whether to recreate stations table

if [ $stations = "create" ]
then
echo "create new stations table from stations.csv"
psql --username "postgres" -d dafni -f /home/sql/stations.sql
else
echo "Load existing stations sql file"
psql --username "postgres" -d dafni -f /home/data/stations/stations_load.sql
fi

psql --username "postgres" -d dafni -c 'create index IF NOT EXISTS idx_stations_location_geom on data.stations using gist ( location_geom );'


f=/home/data/workplace/workplace2011.shp
shp2pgsql -s 27700 -g geom -S -D $f "data"."workplace2011" | psql -U postgres -d dafni

psql --username "postgres" -d dafni -c 'create index IF NOT EXISTS idx_workplace2011_geom on data.workplace2011 using gist ( geom );'
psql --username "postgres" -d dafni -c 'create index IF NOT EXISTS idx_workplace2011_wz on data.workplace2011 using btree ( wz );'

psql --username "postgres" -d dafni -f /home/sql/hhsize.sql

f=/home/data/gb/gb_outline.shp
shp2pgsql -s 27700 -g geom -D $f "data"."gb_outline" | psql -U postgres -d dafni

psql --username "postgres" -d dafni -c 'create index IF NOT EXISTS idx_gb_outline_geom on data.gb_outline using gist ( geom );'


psql --username "postgres" -d dafni -f /home/sql/uplifts.sql

psql --username "postgres" -d dafni -f /home/pgrwrappers/function_pgr_pointtoedgenode.sql
psql --username "postgres" -d dafni -f /home/pgrwrappers/function_bbox_pgr_withpointscost.sql
psql --username "postgres" -d dafni -f /home/pgrwrappers/function_create_pgr_vnodes.sql
psql --username "postgres" -d dafni -f /home/pgrwrappers/function_sdr_crs_pc_nearest_stationswithpoints.sql
psql --username "postgres" -d dafni -f /home/pgrwrappers/function_sdr_pc_station_withpoints.sql
psql --username "postgres" -d dafni -f /home/pgrwrappers/function_sdr_pc_station_withpoints_nobbox.sql

echo "creating centroidnodes table. This will take some time ..."
psql --username "postgres" -d dafni -f /home/sql/centroidnodes.sql







ui/r/prepare_stations.R
# This script creates a series of distance-based service areas for each railway station
# It expects the stations table to be in a schema called 'data'.
# This will take a considerable time to complete (24 hours perhaps).
# Service areas will be used when the nearest stations to each postcode centroid need
# to be generated.

# Preliminaries-----------------------------------------------------------------

args <- commandArgs(trailingOnly = TRUE)
print(args)

setwd("/tmp")

library(sdft)
library(dplyr)
library(tidyr)
library(readr)
library(foreach)
library(stringr)
library(doParallel)
library(keyring)
library(RPostgres)
library(DBI)
library(futile.logger)
library(checkmate)

#key_set_with_value(service = args[1], password = args[2] )

# set up logging

out_path <- getwd()

# delete existing log files
if (file.exists("sdr.log")) {
  file.remove("sdr.log")
}

# delete existing log file
if (file.exists("sa.log")) {
  file.remove("sa.log")
}

threshold <- "DEBUG" # DEBUG, INFO, WARN, ERROR, FATAL
flog.appender(appender.file("sdr.log"))
# set logging level
flog.threshold(threshold)


# capture R errors and warnings to be logged by futile.logger
options(
  showWarnCalls = TRUE,
  showErrorCalls = TRUE,
  show.error.locations = TRUE,
  error = function() {
    flog.error(geterrmessage())
  },
  warning.expression =
    quote({
      if (exists("last.warning", baseenv()) && !is.null(last.warning)) {
        txt = paste0(names(last.warning), collapse = " ")
        flog.warn(txt)
      }
    })
)

# Set up a database connection.
# Using keyring package for storing database password in OS credential store
# to avoid exposing on GitHub. Amend as appropriate.

checkdb <- try(con <-
                 dbConnect(
                   RPostgres::Postgres(),
                   dbname = "dafni",
                   host = "sdft-db",
                   user = "postgres",
                   password = key_get("postgres")
                 ))
if (class(checkdb) == "try-error") {
  stop("Database connection has not been established")
}

# Set up parallel processing
# This is currently used in the sdr_create_service_areas() and
# sdr_generate_choicesets() functions, in a foreach loop.
passwd <- key_get("postgres")

# Number of clusters is total available cores less two.
cl <- makeCluster(detectCores() - 1)
registerDoParallel(cl)
clusterExport(
  cl = cl,
  varlist = c("passwd", "out_path", "threshold"),
  envir = environment()
)

checkcl <- try(clusterEvalQ(cl, {
  library(DBI)
  library(RPostgres)
  library(keyring)
  library(sdft)
  drv <- dbDriver("Postgres")
  con <-
    dbConnect(
      RPostgres::Postgres(),
      host = "sdft-db",
      user = "postgres",
      dbname = "dafni",
      password = passwd
    )
  NULL
}))
if (class(checkcl) == "try-error") {
  stop("clusterEvalQ failed")
}


# create stations dataframe - populate with crscodes and location coordinates
query <- paste0("select crscode, round(st_x(location_geom)) || ',' || round(st_y(location_geom)) as location from data.stations")
stations <- dbGetQuery(con, query)


sdr_create_service_areas(
  con = con,
  out_path = out_path,
  schema = "data",
  df = stations,
  identifier = "crscode",
  table = "stations",
  sa = c(1000, 2000, 3000, 4000, 5000, 10000, 20000, 30000, 40000, 60000, 80000, 105000),
  cost = "len"
)

# 5 minute is to use as fake 60 minute in testing mode

sdr_create_service_areas(
  con,
  out_path,
  schema = "data",
  df = stations,
  identifier = "crscode",
  table = "stations",
  columns = TRUE,
  sa = c(5,60),
  cost = "time"
)



# create spatial indexes for the service areas

sa_names <- c("service_area_1km",
              "service_area_2km",
              "service_area_3km",
              "service_area_4km",
              "service_area_5km",
              "service_area_10km",
              "service_area_20km",
              "service_area_30km",
              "service_area_40km",
              "service_area_60km",
              "service_area_80km",
              "service_area_105km",
              "service_area_60mins")

for (sa_name in sa_names) {
  query <- paste0("create index idx_stations_", sa_name,
                  " on data.stations using gist(", sa_name, ")")
  sdr_dbExecute(con, query)
}

# create 'actual' 60 minute service area for swap between testing modes.

query <- "alter table data.stations add column service_area_60mins_actual geometry(polygon,27700);"
sdr_dbExecute(con, query)

query <- "update data.stations set service_area_60mins_actual = service_area_60mins"
sdr_dbExecute(con, query)
